### DISCUSSION TITLE
How to properly implement Next Script for a Google data structure tag?

### DISCUSSION THREAD

**Discussion by Elindo586** [vercel/next.js#65168] [Help] [answered]
### Summary

Basically, I can't properly view the script loaded in the element window from the Google Developer tools... it gives me a start of the entire JSON file where I have the information, but I can't see much of the script....  So, how can I check if I got it correct??

DETAILS:

I have an index.js file for a blog page loaded with 59 images.

The images are pulled from a JSON file and displayed in chunks of 5 to click on a button and enter an individual video blog.

I want to load an image data structure for this index page with around 59 images.  Because this is a script, I don't need to chunk it in 5... I could just load them all..

According to Google, if I have more than one image, the code would look more like this:

```
<script type="application/ld+json">
    [{
      "@context": "https://schema.org/",
      "@type": "ImageObject",
      "contentUrl": "https://example.com/photos/1x1/black-labrador-puppy.jpg",
      "license": "https://example.com/license",
      "acquireLicensePage": "https://example.com/how-to-use-my-images",
      "creditText": "Labrador PhotoLab",
      "creator": {
        "@type": "Person",
        "name": "Brixton Brownstone"
       },
      "copyrightNotice": "Clara Kent"
    },
   {
      "@context": "https://schema.org/",
      "@type": "ImageObject",
      "contentUrl": "https://example.com/photos/1x1/adult-black-labrador.jpg",
      "license": "https://example.com/license",
      "acquireLicensePage": "https://example.com/how-to-use-my-images",
      "creditText": "Labrador PhotoLab",
      "creator": {
        "@type": "Person",
        "name": "Brixton Brownstone"
       },
      "copyrightNotice": "Clara Kent"
    }]
</script>
```

^^^  To display something like that, I created an element that looks like this:

```
import React from "react";
import Script from "next/script";

const ImageMetadata = (props) => {
  <Script
    id="image-metadata"
    strategy="afterInteractive"
    dangerouslySetInnerHTML={{
      __html: `type="application/ld+json" [{
        "@context": "https://schema.org/",
      "@type": "ImageObject",
      "contentUrl": "${props.twitterImage}",
      "acquireLicensePage": "${props.link}",
      "creator": {
        "@type": "Person",
        "name": "Edgar Lindo"
       },
      "copyrightNotice": "Edgar Lindo",
      }] `,
    }}
  />;
};

export default ImageMetadata;
```

Then... on the page that I want to display it I got this piece of code:

```
 {db.map((item, id) => (
        <div key={id}>
          {console.log(item.twitterImage)}
          <ImageMetadata
            contentUrl={item.twitterImage}
            acquireLicensePage="https://www.tu.biz/video-blog/english"
          />
        </div>
      ))}

```

^^^  
From there, the console.log  logs about 10 images, but why?  I have 59
errr... apparently I am getting all the images in the console.log 

I want to make sure I am creating an array of objects for each image in the script, and not creating an array of images in the `"contentUrl"`  value.

When I try to check it in the element section it looks like this:

![image](https://github.com/vercel/next.js/assets/93617433/0c7bfd27-98bb-49bc-8d29-f210ffddf77f)

**Comment by Elindo586** [comment]
^^^  Now, I am doing something similar for each page with video... 

The video pages only have 1 iframe.... when I check the element section for the script, it also looks like a section of the JSON file.. and  can't see much of the information for the script.

But apparently, for the video pages it seems to work, because Google suddenly indexed a chunk of video pages the day after I implemented the video data structure......  yet, I can't see the code in the element panel.

**Comment by icyJoseph** [comment]
You need to rewrite your script into:

```js

const ImageMetadata = (props) => ( // here you had {
  <Script
    id="image-metadata"

    type="application/ld+json"
    dangerouslySetInnerHTML={{
      __html: `[{
        "@context": "https://schema.org/",
      "@type": "ImageObject",
      "contentUrl": "${props.twitterImage}",
      "acquireLicensePage": "${props.link}",
      "creator": {
        "@type": "Person",
        "name": "Edgar Lindo"
       },
      "copyrightNotice": "Edgar Lindo",
      }] `,
    }}
  />;
); // here you had }
```

That'll create one ld+json script, with one array, with one element object.

**Reply by icyJoseph** [reply]
What you show in the image, `__NEXT_DATA__` that's added by Next.js, and it has nothing to do with your script tags.

**Reply by Elindo586** [reply]
OK, so I changed the type and I put it on top with the id and strategy.

But for some reason I can't locate the script from the Element tab in the developer tools..  Is it there and I just can pin point it?

This would be the page:
https://github.com/Elindo586/nextprojectb/blob/main/pages/video-blog/english/index.js


This is what I see on the element tab:
![image](https://github.com/vercel/next.js/assets/93617433/68744dc3-5627-4951-b45d-6a0860749dce)


While in the individual video pages I can see the added script just fine:

![image](https://github.com/vercel/next.js/assets/93617433/caacdfb2-0f7f-46e9-9def-d952d43ab547)

**Reply by Elindo586** [reply]
BTW... I am console logging the images just fine.  I just can't find the script on the element tab to see how it looks

**Reply by icyJoseph** [reply]
Right, could you also remove the strategy?

**Reply by icyJoseph** [reply]
Ah, I see your error now. You are doing:

```js
export const ImageMetadata = (props) => {
   <Script 
     // Rest of your code
```

Notice that after the arrow you have a `{`, https://github.com/Elindo586/nextprojectb/blob/main/components%2Fdata-structure%2Fimage-metadata.js, and then inside you don't return the Script JSX, so this image metadata component returns undefined.

Change the wrapping `{` `}` to parentheses `(`, `)`

I edited my original answer with what it should look like, using an implicit return as you originally intended

**Reply by Elindo586** [reply]
Man... I need to learn when to use` { }` and when to use `( )`.... that fixed that part...

Now I can see the script in the console.

Why is the console / code only showing 1 item from the `contentUrl={item.twitterImage}` ?

When I console.log the item.twitterImage I can see all of them, but when I input it in the `contentUrl`, then I only see one.  This is the part where I was expecting to see an array of objects with all the different `contentUrl` from the map.

I got this:
```
{db.map((item, id) => (
        <div key={id}>
          {console.log(item.twitterImage)}
          <ImageMetadata
            contentUrl={item.twitterImage}
            acquireLicensePage="https://www.tu.biz/video-blog/english"
          />
        </div>
      ))}

```

And the Element tab is showing this:
![image](https://github.com/vercel/next.js/assets/93617433/12e9bb72-138d-4f89-be51-f2de8791e1fe)

^^ only one object with one `contentUrl` value

**Reply by icyJoseph** [reply]
Errrr twitterImage is a string in your db, do you mean that you want to map over the db collecting the twitterImage per entry and then using that as the input to your ImageMetadata component?

**Reply by icyJoseph** [reply]
Like this

```js
          <ImageMetadata
            contentUrl={db.map(item => item.twitterImage)}
         acquireLicensePage="https://www.tu.biz/video-blog/english"
          />
```

**Reply by Elindo586** [reply]
OK... but that inputs a string of image addresses in the `contentUrl`

According to Google, I need to create an array of objects, in this case, each object would have a different value for `contentUrl`, coming from the map of `item.twitterImage`

Would end up some kind of like this:

```
<script type="application/ld+json">
    [{
      "@context": "https://schema.org/",
      "@type": "ImageObject",
      "contentUrl": "https://example.com/here-goes-my-1.jpg",
      "license": "https://example.com/license",
      "acquireLicensePage": "https://example.com/how-to-use-my-images",
      "creditText": "Labrador PhotoLab",
      "creator": {
        "@type": "Person",
        "name": "Brixton Brownstone"
       },
      "copyrightNotice": "Clara Kent"
    },
   {
      "@context": "https://schema.org/",
      "@type": "ImageObject",
      "contentUrl": "https://example.com/here-goes-my-2.jpg",
      "license": "https://example.com/license",
      "acquireLicensePage": "https://example.com/how-to-use-my-images",
      "creditText": "Labrador PhotoLab",
      "creator": {
        "@type": "Person",
        "name": "Brixton Brownstone"
       },
      "copyrightNotice": "Clara Kent"
    }]
</script>

```

^^^  I would end up with one script containing about 59 objects, one object per image address.

**Reply by icyJoseph** [reply]
You almost have it... So your image metadata component now has the list, so within that component, you map over props.twitterImage, building the list of objects you need, and then... I don't want to spoil this learning opportunity, but you are really close

**Reply by Elindo586** [reply]
Ok... give sometime.... :)

**Reply by icyJoseph** [reply]
Btw I would say that the question has been resolved, but feel free to write further more

**Reply by Elindo586** [reply]
Ok..

I went to the component and decided to do it at all the component... still only showing 1  object ...

I tried mapping the `db` and put it in a new variable, but it gave me the same result. 

 the code I have is:

```

import React, { useState, useEffect } from "react";
import Script from "next/script";
import db from "../../utils/blogs-front/english/blog-english";

const ImageMetadata = (props) => {
  const [index, setIndex] = useState(0);

  useEffect(() => {
    setIndex((prevIndex) => (prevIndex + 1) % db.length);
  }, []);

  const contentUrl = db[index].twitterImage;

  return (
    <Script
      id="image-metadata"
      type="application/ld+json"
      dangerouslySetInnerHTML={{
        __html: `[{
        "@context": "https://schema.org/",
      "@type": "ImageObject",
      "contentUrl": "${contentUrl}",
      "acquireLicensePage": "${props.acquireLicensePage}",
      "creator": {
        "@type": "Person",
        "name": "Edgar Lindo"
       },
      "copyrightNotice": "Edgar Lindo",
      }] `,
      }}
    />
  );
};

export default ImageMetadata;

```

The result still just one object:

![image](https://github.com/vercel/next.js/assets/93617433/3a44b185-0c08-47e9-9ee4-c1a0b81860e3)

**Reply by icyJoseph** [reply, marked as answer]
I see, that was not quite what you needed.

```js
const ImageMetadata = (props) => {
  const content = db.map((item) => ({
    '@context': 'https://schema.org/',
    '@type': 'ImageObject',
    contentUrl: item.twitterImage,
    acquireLicensePage: `${props.acquireLicensePage}`,
    creator: {
      '@type': 'Person',
      name: 'Edgar Lindo',
    },
    copyrightNotice: 'Edgar Lindo',
  }));

  return (
    <Script
      id="image-metadata"
      type="application/ld+json"
      dangerouslySetInnerHTML={{
        __html: JSON.stringify(content, null, 2),
      }}
    />
  );
};
```

**Reply by Elindo586** [reply]
I would have never gotten that...  map over an item, plus include the parts of the object you need, and then JSON.stringify it on the script..

Normally, I would just map over something, and put it on inside a table tags, and it fills the table by itself... but yes, we have something different here...


Thank you much...
